OpenRA.Mods.Common/Traits/Render/WithAttackAnimation.cs

```diff
@@ -34,68 +34,87 @@ public class WithAttackAnimationInfo : ITraitInfo, Requires<WithSpriteBodyInfo>,
 		[Desc("Should the animation be delayed relative to preparation or actual attack?")]
 		public readonly AttackDelayType DelayRelativeTo = AttackDelayType.Preparation;
 
-		public object Create(ActorInitializer init) { return new WithAttackAnimation(init, this); }
+		[Desc("Which sprite body to modify.")]
+		public readonly string BodyName = "body";
+
+		public override object Create(ActorInitializer init) { return new WithAttackAnimation(init, this); }
+
+		public override void RulesetLoaded(Ruleset rules, ActorInfo ai)
+		{
+			var matches = ai.TraitInfos<WithSpriteBodyInfo>().Count(w => BodyName == w.Name);
+			if (matches != 1)
+				throw new YamlException("WithAttackAnimation needs exactly one sprite body with matching name.");
+
+			base.RulesetLoaded(rules, ai);
+		}
 	}
 
-	public class WithAttackAnimation : ITick, INotifyAttack
+	public class WithAttackAnimation : ConditionalTrait<WithAttackAnimationInfo>, ITick, INotifyAttack
 	{
-		readonly WithAttackAnimationInfo info;
 		readonly AttackBase attack;
 		readonly Armament armament;
 		readonly WithSpriteBody wsb;
+		readonly bool noAimOrReloadAnim;
 
 		int tick;
+		bool attackAnimPlaying;
 
 		public WithAttackAnimation(ActorInitializer init, WithAttackAnimationInfo info)
+			: base(info)
 		{
-			this.info = info;
 			attack = init.Self.Trait<AttackBase>();
 			armament = init.Self.TraitsImplementing<Armament>()
-				.Single(a => a.Info.Name == info.Armament);
-			wsb = init.Self.Trait<WithSpriteBody>();
+				.Single(a => a.Info.Name == Info.Armament);
+			wsb = init.Self.TraitsImplementing<WithSpriteBody>().First(w => Info.BodyName == w.Info.Name);
```

Comment:
minor nit to ignore: for consistency it could be First(w => **w**.Info.Name == Info.BodyName) (also in RulesetLoaded). Not changed by this PR but I assume .Single could be replaced by First which is faster. We check for single WithSpriteBodyInfo in RulesetLoaded so why should we rely on exception here for Armament?